File: examples/androidJavaSDKExample/app/src/main/java/com/llamamobile/sdkexample/ChatAdapter.java
package com.llamamobile.sdkexample;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;

public class ChatAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int VIEW_TYPE_USER = 1;
    private static final int VIEW_TYPE_ASSISTANT = 2;

    private final List<Message> messages;

    public ChatAdapter(List<Message> messages) {
        this.messages = messages;
    }

    @Override
    public int getItemViewType(int position) {
        String role = messages.get(position).getRole();
        if (Message.ROLE_USER.equals(role)) {
            return VIEW_TYPE_USER;
        } else if (Message.ROLE_ASSISTANT.equals(role)) {
            return VIEW_TYPE_ASSISTANT;
        } else {
            return VIEW_TYPE_ASSISTANT; // Default to assistant
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        switch (viewType) {
            case VIEW_TYPE_USER:
                View userView = inflater.inflate(R.layout.item_message_user, parent, false);
                return new UserMessageViewHolder(userView);
            case VIEW_TYPE_ASSISTANT:
                View assistantView = inflater.inflate(R.layout.item_message_assistant, parent, false);
                return new AssistantMessageViewHolder(assistantView);
            default:
                throw new IllegalArgumentException("Unknown view type: " + viewType);
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        Message message = messages.get(position);
        if (holder instanceof UserMessageViewHolder) {
            ((UserMessageViewHolder) holder).bind(message);
        } else if (holder instanceof AssistantMessageViewHolder) {
            ((AssistantMessageViewHolder) holder).bind(message);
        }
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    // User message view holder
    public static class UserMessageViewHolder extends RecyclerView.ViewHolder {
        private final TextView messageText;
        private final TextView messageSender;

        public UserMessageViewHolder(View itemView) {
            super(itemView);
            messageText = itemView.findViewById(R.id.userMessageText);
            messageSender = itemView.findViewById(R.id.userMessageSender);
        }

        public void bind(Message message) {
            messageText.setText(message.getText());
            messageSender.setText("You");
        }
    }

    // Assistant message view holder
    public static class AssistantMessageViewHolder extends RecyclerView.ViewHolder {
        private final TextView messageText;
        private final TextView messageSender;

        public AssistantMessageViewHolder(View itemView) {
            super(itemView);
            messageText = itemView.findViewById(R.id.assistantMessageText);
            messageSender = itemView.findViewById(R.id.assistantMessageSender);
        }

        public void bind(Message message) {
            messageText.setText(message.getText());
            messageSender.setText("Llama");
        }
    }
}

File: examples/androidSDKExample/app/src/main/java/com/llamamobile/sdkexample/ChatAdapter.kt
package com.llamamobile.sdkexample

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

class ChatAdapter(private val messages: List<Message>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val VIEW_TYPE_USER = 1
        private const val VIEW_TYPE_ASSISTANT = 2
    }

    override fun getItemViewType(position: Int): Int {
        return when (messages[position].role) {
            Message.ROLE_USER -> VIEW_TYPE_USER
            Message.ROLE_ASSISTANT -> VIEW_TYPE_ASSISTANT
            else -> VIEW_TYPE_ASSISTANT
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            VIEW_TYPE_USER -> {
                val view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_message_user, parent, false)
                UserMessageViewHolder(view)
            }
            VIEW_TYPE_ASSISTANT -> {
                val view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_message_assistant, parent, false)
                AssistantMessageViewHolder(view)
            }
            else -> throw IllegalArgumentException("Unknown view type: $viewType")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val message = messages[position]
        when (holder) {
            is UserMessageViewHolder -> holder.bind(message)
            is AssistantMessageViewHolder -> holder.bind(message)
        }
    }

    override fun getItemCount(): Int {
        return messages.size
    }

    // User message view holder
    class UserMessageViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val messageText: TextView = itemView.findViewById(R.id.userMessageText)
        private val messageSender: TextView = itemView.findViewById(R.id.userMessageSender)

        fun bind(message: Message) {
            messageText.text = message.text
            messageSender.text = "You"
        }
    }

    // Assistant message view holder
    class AssistantMessageViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val messageText: TextView = itemView.findViewById(R.id.assistantMessageText)
        private val messageSender: TextView = itemView.findViewById(R.id.assistantMessageSender)

        fun bind(message: Message) {
            messageText.text = message.text
            messageSender.text = "Llama"
        }
    }
}
